### ISSUE TITLE
invalid-argument-type when narrowing a sum type with isintance

### ISSUE THREAD

**Issue by pcastellazzi** [astral-sh/ty#2742]
### Summary

When running `ty` on the example code (see below) with:

```bash
$ uvx --python=3.14 --from=ty@0.0.15 ty check example.py
```

I got the following error:

```plain
error[invalid-argument-type]: Argument to bound method `append` is incorrect
  --> example.py:27:33
   |
25 |             state = option(index, text)
26 |             if isinstance(state, Failure):
27 |                 failures.append(state.error)
   |                                 ^^^^^^^^^^^ Expected `F@choice1`, found `object`
28 |             else:
29 |                 return state
   |
info: Method defined here
    --> stdlib/builtins.pyi:2843:9
     |
2841 |         """Return a shallow copy of the list."""
2842 |
2843 |     def append(self, object: _T, /) -> None:
     |         ^^^^^^       ---------- Parameter declared here
2844 |         """Append object to the end of the list."""
     |
info: rule `invalid-argument-type` is enabled by default
```

Relevant code: https://play.ty.dev/588d142d-5241-4854-8e6b-18274d20946b

I have a few questions:
-  The two implementations should be equivalent, but `ty` is reporting `state.error` to be of type `object `. Since there is not such thing as `isinstance(state, Failure[F])` this is kinda true. Not sure if this is expected behavior and I should fix the code or an error. `state.error` is marked as `@todo` on the implementation using match, hence the question.
- I can't reproduce the problem in the playground. Whatever version of `ty` is running works correctly. is this a regression?

I tried different combinations of Python and Ty without luck, using:
```bash
for py in $(seq 12 14); do
    for ty in $(seq 10 15); do
        uvx --python="3.$py" --from="ty@0.0.$ty" ty check example.py && break
    done
done
```

### Version

0.0.15

**Comment by carljm** [comment]
Thanks for the report! Playground always runs the latest main branch (given a few minutes to deploy after each commit lands), but I think the key here is Python version. If we [modify the playground to check with Python 3.12](https://play.ty.dev/d756f74d-74e9-4559-94e7-462dfac46c35) (see the `ty.json` tab), we see the error you report. On Python 3.13 or 3.14: no error.

The reason for this discrepancy is that Python 3.13 added a `__replace__` method to dataclasses, which makes frozen generic dataclasses invariant instead of covariant, according to variance inference (since the `__replace__` method signature has the type variable in contravariant - input - position in a method.) This is kind of unfortunate in general, since covariance is often useful, but in this case it actually fixes the problem, because the invariance means we no consider it possible for an object to be both a `Success[S]` and a `Failure`.

With covariance, though, we do consider it to be possible, due to multiple inheritance. And that's the cause of the issue. As you said, "there is not such thing as `isinstance(state, Failure[F])`", so with covariant typing we consider `isinstance(state, Failure)` to establish a type of `Failure[object]`, and if you inspect the type of `state` after that check, it is `Failure[F@choice1] | (Success[S@choice1] & Failure[object])` -- that is, it's either a `Failure[F]` or it's an intersection of both `Success[S]` and `Failure[object]`.

The easy fix here, presuming you don't use subclasses of `Failure` and `Success`, is to make them final: that [fixes the problem in 3.12](https://play.ty.dev/21f24db8-a365-400b-ada5-f58a89debe60) also.

If you do want subclassing of `Success` and `Failure`, then it becomes a bit harder to deal with this in a nice way. More discussion in #1578

**Comment by carljm** [comment]
Oh, and the Todo type on the match version is just incomplete implementation of match statements in ty; see #887

**Comment by carljm** [comment]
I guess the last question here is why this didn't uncover the version-dependence for you:

> I tried different combinations of Python and Ty without luck, using:

> for py in $(seq 12 14); do
    for ty in $(seq 10 15); do
        uvx --python="3.$py" --from="ty@0.0.$ty" ty check example.py && break
    done
done

I suspect the reason is that you have a `pyproject.toml` with a `requires-python` version of 3.12 in the same directory, or a parent directory, of your `example.py`? ty considers that a higher-priority signal for Python version to check with than the Python version of the venv used. If you want to be sure ty will check with a certain Python version, pass `--python-version` option to ty itself.

**Comment by pcastellazzi** [comment]
@carljm Thank you very much for your detailed analysis.

> The easy fix here, presuming you don't use subclasses of `Failure` and `Success`, is to make them final: that [fixes the problem in 3.12](https://play.ty.dev/21f24db8-a365-400b-ada5-f58a89debe60) also.

For this particular case `final` is fine. Good catch.

> Oh, and the Todo type on the match version is just incomplete implementation of match statements in ty; see [#887](https://github.com/astral-sh/ty/issues/887)

I meant, since the match implementation is not complete, i could not compare the two implementations. 

> I suspect the reason is that you have a `pyproject.toml` with a `requires-python` version of 3.12 in the same directory, or a parent directory, of your `example.py`? ty considers that a higher-priority signal for Python version to check with than the Python version of the venv used. If you want to be sure ty will check with a certain Python version, pass `--python-version` option to ty itself.

That's indeed the case. Again, good catch, and also good to know. I wasn't aware that was the case.
